apps/designer/app/designer/features/workspace/canvas-tools/text-toolbar/text-toolbar.tsx

```diff
@@ -44,51 +41,52 @@ export const useSubscribeTextToolbar = () => {
   useSubscribe("hideTextToolbar", () => setTextToolbar(undefined));
 };
 
-const getPlacement = ({
-  toolbarRect,
-  selectionRect,
-}: {
-  toolbarRect?: DOMRect;
-  selectionRect: DOMRect;
-}) => {
-  let align = "top";
-  let left = selectionRect.x + selectionRect.width / 2;
-  // We measure the size in a hidden state after we render the menu,
-  // then show it
-  let visibility = "hidden";
-  if (toolbarRect !== undefined) {
-    visibility = "visible";
-    // Prevent going further than left 0
-    left = Math.max(left, toolbarRect.width / 2);
-    // Prevent going further than window width
-    left = Math.min(left, window.innerWidth - toolbarRect.width / 2);
-    align = selectionRect.y > toolbarRect.height ? "top" : "bottom";
-  }
-
-  const marginBottom = align === "bottom" ? "-$5" : 0;
-  const marginTop = align === "bottom" ? 0 : "-$5";
-  const transform = "translateX(-50%)";
-  const top =
-    align === "top"
-      ? Math.max(selectionRect.y - selectionRect.height, 0)
-      : Math.max(selectionRect.y + selectionRect.height);
-
-  return { top, left, marginBottom, marginTop, transform, visibility };
-};
-
 const onClickPreventDefault: MouseEventHandler<HTMLDivElement> = (event) => {
   event.preventDefault();
   event.stopPropagation();
 };
 
+const computeAbsoluteRect = (parent: DOMRect, rel: DOMRect) => {
+  return {
+    x: parent.x + rel.x,
+    y: parent.y + rel.y,
+    width: rel.width,
+    height: rel.height,
+    top: parent.top + rel.top,
+    left: parent.left + rel.left,
+    bottom: parent.bottom - rel.bottom,
+    right: parent.right - rel.right,
+  };
+};
+
 type ToolbarProps = {
-  css?: CSS;
-  rootRef: React.Ref<HTMLDivElement>;
   state: TextToolbarState;
   onToggle: (value: Format) => void;
 };
 
-const Toolbar = ({ css, rootRef, state, onToggle }: ToolbarProps) => {
+const Toolbar = ({ state, onToggle }: ToolbarProps) => {
+  const rootRef = useRef<HTMLDivElement | null>(null);
+
+  useEffect(() => {
+    if (rootRef.current && rootRef.current.parentElement) {
```

Comment:
ts would understand that bot non null if single check 
`if (rootRef.current?.parentElement)`